### DISCUSSION TITLE
Ideas for debugging realtime events?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#1076] [Q&A]
I'm hitting cases where it seems like record updates are not triggering realtime events. Here's the basic setup:

* PocketBase JS running on server as admin user
* PocketBase JS running on web frontend as normal user

```ts
// Frontend
const job = await client.collection('jobs').create({status: 'new'})
const unsub = await client.collection('jobs').subscribe(job.id, console.log)

// I need this because it's possible that the job may have completed before
// subscribe() even takes effect. I wish there were a createAndSubscribe()
// method that guaranteed no realtime updates would be missed
console.log(await client.collection('jobs').getOne(job.id))

// Backend
const unsub = await client.collection('jobs').subscribe('*', async e=>{
   if(e.action == 'create') {
      await client.collection('jobs').update(e.record.id, {status: 'started'})
      setTimeout(async ()=>{
         await client.collection('jobs').update(e.record.id, {status: 'finished'})
      }, Math.random()*1000)
   }
})
```

Questions:

1. Is there a way to have PocketBase emit debugging output every time a SSE update is sent?
2. What's the best way to debug SSE traffic in the browser?
3. What's the best way to debug SSE traffic in a node client?

**Comment by ganigeorgiev** [comment]
>  I wish there were a createAndSubscribe() method that guaranteed no realtime updates would be missed

I haven't tested it, but you should be able to set a custom job id (it needs to be 15 chars string) and subscribe to it upfront, aka:

```js
const jobId = "abc123";

const unsub = await client.collection('jobs').subscribe(jobId, console.log)

const job = await client.collection('jobs').create({id: jobId, status: 'new'})
```

---

> Is there a way to have PocketBase emit debugging output every time a SSE update is sent?

I'm not sure that I understand what you are looking for. What should be the expected result if you have 100 users subscribed to the same topic and a record event occur - print 100 messages in the console or just print a message that a record event has occur? If the latter then you can attach to the `OnModel*` hooks (the realtime api do exactly that).

---

> What's the best way to debug SSE traffic in the browser?

You can inspect all SSE event messages sent to the client in the browsers "Inspect tool > Network tab" and find the `GET /api/realtime` request.

---

> What's the best way to debug SSE traffic in a node client?

The SSE connections are the same as any other HTTP connections, so technically any network traffic monitoring tool should be able to help with the debug (eg. wireshark). So far I mainly used `console.log` in the callback when I had to test something quick in node.

---

In general, if you are not receiving events:

- start the PocketBase executable with `--debug` (on error we may log in the cmd the reason why something has failed)

- make sure that, if you are not authenticated as admin, your user/client has List (for `*`) or View (for single record) API rules access
